### PR TITLE
[Serve] Support 'extra' in logging

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

```
@serve.deployment
class Test:
    def __init__(self):
        self.logger = logging.getLogger("ray.serve")
        #pass

    def __call__(self):
        self.logger.info('"123"', extra={"ray_serve_extra_fields": {"my_key": "my_value"}})
```

```
{"levelname": "INFO", "asctime": "2023-11-06 11:44:42,827", "deployment": "Test", "replica": "default#Test#UpEyII", "request_id": "472a1c26-2dfb-4e80-b819-198bca6bc96e", "route": "/", "my_key": "my_value", "application": "default", "message": "test_log_print.py:20 - 123"}
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/logging_utils.py [hunk 1]
```diff
@@ -74,6 +74,11 @@ def format(self, record: logging.LogRecord) -> str:
             ]
         if SERVE_LOG_ROUTE in record.__dict__:
             record_format[SERVE_LOG_ROUTE] = SERVE_LOG_RECORD_FORMAT[SERVE_LOG_ROUTE]
+
+        if "_extra" in record.__dict__ and record.__dict__["_extra"]:
+            for key in record.__dict__["_extra"]:
+                record_format[key] = f"%({key})s"
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/logging_utils.py:80]: Could you log a warning if the `key` already exists in `record_format`. That would help prevent users from overwriting existing keys.

File: python/ray/serve/_private/logging_utils.py [hunk 2]
```diff
@@ -161,6 +166,33 @@ def get_component_logger_file_path() -> Optional[str]:
...
+        self,
+        name,
+        level,
+        fn,
+        lno,
+        msg,
+        args,
+        exc_info,
+        func=func,
+        extra=extra,
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/logging_utils.py:189]: What does passing in `extra` do here? Will the `extra` fields show up twice in the log record– once when we explicitly copy them using the `_extra` hack, and once when `original_makeRecord` parses them?

File: python/ray/serve/_private/logging_utils.py [hunk 3]
```diff
@@ -161,6 +166,33 @@ def get_component_logger_file_path() -> Optional[str]:
...
+        sinfo=sinfo,
+    )
+    record._extra = extra
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/logging_utils.py:192]: Should we `deepcopy` `extra` here in case either Serve or the `logging` package mutates the `extra` dict in the future?

File: python/ray/serve/_private/logging_utils.py (hunk 1 above)
Comment (sihanwang41) [reply python/ray/serve/_private/logging_utils.py]: not applied, updated pr without using "_extra"

File: python/ray/serve/_private/logging_utils.py (hunk 2 above)
Comment (sihanwang41) [reply python/ray/serve/_private/logging_utils.py]: not applied, updated pr without using "_extra"

File: python/ray/serve/_private/logging_utils.py (hunk 3 above)
Comment (sihanwang41) [reply python/ray/serve/_private/logging_utils.py]: not applied, updated pr without using "_extra"

File: python/ray/serve/_private/logging_utils.py [hunk 4]
```diff
@@ -74,12 +75,20 @@ def format(self, record: logging.LogRecord) -> str:
...
             record_format[SERVE_LOG_ROUTE] = SERVE_LOG_RECORD_FORMAT[SERVE_LOG_ROUTE]
+
         if SERVE_LOG_APPLICATION in record.__dict__:
             record_format[SERVE_LOG_APPLICATION] = SERVE_LOG_RECORD_FORMAT[
                 SERVE_LOG_APPLICATION
             ]
 
-        record_format[SERVE_LOG_MESSAGE] = SERVE_LOG_RECORD_FORMAT[SERVE_LOG_MESSAGE]
+        if SERVE_LOG_EXTRA_FIELDS in record.__dict__:
+            for k, v in record.__dict__[SERVE_LOG_EXTRA_FIELDS].items():
```
Comment (GeneDer) [inline python/ray/serve/_private/logging_utils.py:85]: Are there mechanisms to ensure this is always a dictionary?

File: python/ray/serve/_private/logging_utils.py [hunk 5]
```diff
@@ -74,12 +75,20 @@ def format(self, record: logging.LogRecord) -> str:
...
+                record_format[k] = v
```
Comment (GeneDer) [inline python/ray/serve/_private/logging_utils.py:86]: Things already in `record_format` could be override by this. Is that intended? Or maybe we need a validation for it?

File: python/ray/serve/tests/test_logging.py [hunk 6]
```diff
@@ -280,6 +281,29 @@ def check_log_file(log_file: str, expected_regex: list):
         check_log_file(resp2["log_file"], user_class_method_log_regexes)
 
 
+def test_extra_field(serve_and_ray_shutdown):
+    """Test ray serve extra logging"""
+    ray.init(runtime_env={"env_vars": {"RAY_SERVE_ENABLE_JSON_LOGGING": "1"}})
```
Comment (edoakes) [inline python/ray/serve/tests/test_logging.py:286]: What happens if `RAY_SERVE_ENABLE_JSON_LOGGING` is not enabled and someone uses the extra fields? they're ignored?

**Review by edoakes — COMMENTED** [review]
Looks good, add a section to the monitoring docs?

File: python/ray/serve/_private/logging_utils.py [hunk 7]
```diff
@@ -74,12 +75,20 @@ def format(self, record: logging.LogRecord) -> str:
...
+
+        if SERVE_LOG_MESSAGE in record.__dict__:
+            record_format[SERVE_LOG_MESSAGE] = (
+                SERVE_LOG_RECORD_FORMAT[SERVE_LOG_MESSAGE] % record.__dict__
+            )
```
Comment (zcin) [inline python/ray/serve/_private/logging_utils.py:91]: Why is this changed? I thought the entries of `SERVE_LOG_RECORD_FORMAT` are not "filled in" with values yet

File: python/ray/serve/tests/test_logging.py (hunk 6 above)
Comment (sihanwang41) [reply python/ray/serve/tests/test_logging.py:286]: nothing happens, we just ignore.

this aligns with the current logging lib behavior. 
```
The fourth keyword argument is extra which can be used to pass a dictionary which is used to populate the __dict__ of the [LogRecord](https://docs.python.org/3/library/logging.html#logging.LogRecord) created for the logging event with user-defined attributes
```

https://docs.python.org/3/library/logging.html#logging.debug

**Comment by sihanwang41** [comment]
> Looks good, add a section to the monitoring docs?

Hi Ed, I am planning to do doc in separate pr.

File: python/ray/serve/_private/logging_utils.py (hunk 7 above)
Comment (sihanwang41) [reply python/ray/serve/_private/logging_utils.py:91]: this is to early-fill in the values and jsonize the value to avoid non-jsonizable value in the log content.

File: python/ray/serve/_private/logging_utils.py (hunk 4 above)
Comment (sihanwang41) [reply python/ray/serve/_private/logging_utils.py:85]: add a check and raise error.

File: python/ray/serve/_private/logging_utils.py (hunk 5 above)
Comment (sihanwang41) [reply python/ray/serve/_private/logging_utils.py:86]: nice catch, added check.

File: python/ray/serve/_private/logging_utils.py [hunk 8]
```diff
@@ -74,12 +75,27 @@ def format(self, record: logging.LogRecord) -> str:
...
+
+        if SERVE_LOG_EXTRA_FIELDS in record.__dict__:
+            if not isinstance(record.__dict__[SERVE_LOG_EXTRA_FIELDS], dict):
+                raise ValueError(
+                    f"Expected a dictionary passing into {SERVE_LOG_EXTRA_FIELDS}, "
+                    f"but got {type(record.__dict__[SERVE_LOG_EXTRA_FIELDS])}"
+                )
+            for k, v in record.__dict__[SERVE_LOG_EXTRA_FIELDS].items():
+                if k in record_format:
+                    raise KeyError(f"Found duplicated key in the log record: {k}")
```
Comment (GeneDer) [inline python/ray/serve/_private/logging_utils.py:97]: Nitpick (non-blocker): would be nice to add a unit test for this and when `record.__dict__[SERVE_LOG_EXTRA_FIELDS]` is not a dictionary 🙂

**Review by GeneDer — APPROVED** [review]
Thanks for addressing the comments!

**Comment by edoakes** [comment]
Code LGTM, planning to add docs here or in follow-up>?

**Comment by sihanwang41** [comment]
> Code LGTM, planning to add docs here or in follow-up>?

haha, separate pr. https://github.com/ray-project/ray/pull/40985#issuecomment-1802258544  :)

**Review by edoakes — APPROVED** [review]

**Review by zcin — APPROVED** [review]

File: python/ray/serve/_private/logging_utils.py (hunk 8 above)
Comment (sihanwang41) [reply python/ray/serve/_private/logging_utils.py:97]: done

**Comment by sihanwang41** [comment]
@architkulkarni ready to merge!
